cartoframes/datasets.py

```diff
@@ -0,0 +1,252 @@
+import binascii as ba
+from warnings import warn
+
+import pandas as pd
+from carto.exceptions import CartoException
+
+
+class Dataset(object):
+    SUPPORTED_GEOM_COL_NAMES = ['geom', 'the_geom', 'geometry']
+    UTIL_COLS = SUPPORTED_GEOM_COL_NAMES + ['the_geom_webmercator', 'cartodb_id']
+
+    FAIL = 'fail'
+    REPLACE = 'replace'
+    APPEND = 'append'
+
+    def __init__(self, carto_context, table_name, df=None):
+        self.cc = carto_context
+        self.table_name = norm_colname(table_name)
+        self.df = df
+        warn('Table will be named `{}`'.format(table_name))
+
+    def upload(self, with_lonlat=None, if_exists='fail'):
+        if self.df is None:
+            raise ValueError('You have to create a `Dataset` with a pandas DataFrame in order to upload it to CARTO')
+
+        if not self.exists():
+            self._create_table(with_lonlat)
+        else:
+            if if_exists == Dataset.FAIL:
+                raise ValueError(('Table with name {table_name} already exists in CARTO.'
+                                  ' Please choose a different `table_name` or use'
+                                  ' if_exists="replace" to overwrite it').format(table_name=self.table_name))
+            elif if_exists == Dataset.REPLACE:
+                self._create_table(with_lonlat)
+
+        self._copyfrom(with_lonlat)
+
+        return self
+
+    def exists(self):
+        """Checks to see if table exists"""
+        try:
+            self.cc.sql_client.send(
+                'EXPLAIN SELECT * FROM "{table_name}"'.format(
+                    table_name=self.table_name),
+                do_post=False)
+            return True
+        except CartoException as err:
+            # If table doesn't exist, we get an error from the SQL API
+            self.cc._debug_print(err=err)
+            return False
+
+    def _create_table(self, with_lonlat=None):
+        job = self.cc.batch_sql_client \
+                  .create_and_wait_for_completion(
+                      '''BEGIN; {drop}; {create}; {cartodbfy}; COMMIT;'''
+                      .format(drop=self._drop_table_query(),
+                              create=self._create_table_query(with_lonlat),
+                              cartodbfy=self._cartodbfy_query()))
+
+        if job['status'] != 'done':
+            raise CartoException('Cannot create table: {}.'.format(job['failed_reason']))
+
+    def _cartodbfy_query(self):
+        return "SELECT CDB_CartodbfyTable('{org}', '{table_name}')" \
+            .format(org=(self.cc.creds.username() if self.cc.is_org else 'public'),
+                    table_name=self.table_name)
+
+    def _copyfrom(self, with_lonlat=None):
+        util_cols = Dataset.UTIL_COLS
```

Comment:
Why duplicate this instead of just using the constant in the comprehension below?

Reply:
No reason actually, is it better one thing or the other?

Reply:
I will use the constant too. And the same with `geom_cols`